internal/sshtunneler/tunneler.go

```diff
@@ -0,0 +1,278 @@
+	privKey, err := ssh.ED25519()
+	if err != nil {
+		return nil, nil, errors.Annotate(err, "failed to generate key")
+	}
+
+	sshPrivateKey, err := gossh.NewSignerFromKey(privKey)
+	if err != nil {
+		return nil, nil, errors.NotValidf("private key")
+	}
+
+	return sshPrivateKey, sshPrivateKey.PublicKey(), nil
+}
+
+// RequestTunnel requests a tunnel to a model specific unit.
+//
+// The returned tunnelRequest should be used to wait for the tunnel to be established.
+// See Wait() for more information.
+func (tt *Tracker) RequestTunnel(req RequestArgs) (*Request, error) {
+	tunnelID, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	// We use the same expiry for the password and the state entry.
+	// The state's expiry is used to clean up any dangling requests.
+	// The password expiry is used to invalidate old passwords.
+	expiry := tt.clock.Now().Add(maxTimeout)
+
+	password, err := tt.authn.generatePassword(tunnelID.String(), expiry)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKey, publicKey, err := tt.generateEphemeralSSHKey()
+	if err != nil {
+		return nil, err
+	}
+
+	args := state.SSHConnRequestArg{
+		TunnelID:            tunnelID.String(),
+		ModelUUID:           req.ModelUUID,
+		MachineId:           req.MachineID,
+		Expires:             expiry,
+		Username:            reverseTunnelUser,
+		Password:            password,
+		ControllerAddresses: tt.controller.Addresses(),
+		UnitPort:            22,
+		EphemeralPublicKey:  publicKey.Marshal(),
+	}
+
+	err = tt.state.InsertSSHConnRequest(args)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanup := func() {
+		tt.delete(tunnelID.String())
+	}
+	// Make sure to use an unbuffered channel to ensure someone always
+	// has responsibility of the connection passed around.
+	tunnelReq := &Request{
+		recv:       make(chan (net.Conn)),
+		privateKey: privateKey,
+		cleanup:    cleanup,
+		dialer:     tt.dialer,
+	}
+
+	tt.add(tunnelID.String(), tunnelReq)
+
+	return tunnelReq, nil
+}
+
+func (tt *Tracker) add(tunnelID string, req *Request) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	tt.tracker[tunnelID] = req
+}
+
+func (tt *Tracker) get(tunnelID string) (*Request, bool) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	req, ok := tt.tracker[tunnelID]
+	return req, ok
+}
+
+func (tt *Tracker) delete(tunnelID string) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	delete(tt.tracker, tunnelID)
+}
+
+// AuthenticateTunnel authenticates an SSH request for a tunnel.
+//
+// An SSH server is expected to call this method to validate that
+// the connection is a valid tunnel request.
+//
+// If the request is valid, the provided tunnelID should be
+// stored and provided alongside the network connection to PushTunnel.
+func (tt *Tracker) AuthenticateTunnel(username, password string) (tunnelID string, err error) {
+	if username != reverseTunnelUser {
+		return "", errors.New("invalid username")
+	}
+
+	return tt.authn.validatePassword(password)
```

Comment:
Can we check that the tunnel still exists here? Fine to check the password is a valid token issued by us, but also good to check earlier the tunnel id in the claims is still waiting.